pallene/constant_propagation.lua

```diff
@@ -16,128 +16,185 @@ local function is_constant_value(v)
+                        next_f.is_upvalue_constant[u_id] = n_writes[value.id] == 1
+                            and f_data.locvar_constant_init[value.id]
+                    elseif value._tag == "ir.Value.Upvalue" then
+                        next_f.is_upvalue_constant[u_id] = f_data.is_upvalue_constant[value.id]
+                    end
+                end
+
             else
-                -- skip
+                local dsts = ir.get_dsts(cmd)
+                for _, dst_id in ipairs(dsts) do
+                    n_writes[dst_id] = n_writes[dst_id] + 1
+                end
             end
         end
     end
 
-    -- 3) Find out which constant globals should be propagated, and which unused constant globals
-    -- should be simply eliminated.
-
-    local is_exported = {}
-    for _, g_id in ipairs(module.exported_globals) do
-        is_exported[g_id] = true
-    end
+    -- 3) Remove propagated upvalues from the capture list.
+    for _, func in ipairs(module.functions) do
+        for cmd in ir.iter(func.body) do
+            if cmd._tag == "ir.Cmd.SetUpvalues" then
+                local next_f   = assert(data_of_func[cmd.f_id])
+                local ir_func  = module.functions[cmd.f_id]
+                local new_u_id = next_f.new_upvalue_id
+
+                local new_srcs = {}
+                for u_id, value in ipairs(cmd.srcs) do
+                    if not next_f.is_upvalue_constant[u_id] then
+                        table.insert(new_srcs, value)
+                        new_u_id[u_id] = #new_srcs
+                    end
+                end
 
-    local new_globals = {}
-    local new_global_id = {} -- { g_id => g_id? }
-    for i = 1, n_globals do
-        if constant_initializer[i] and not is_exported[i] and (n_reads[i] == 0 or n_writes[i] == 1) then
-            new_global_id[i] = false
-        else
-            table.insert(new_globals, module.globals[i])
-            new_global_id[i] = #new_globals
+                cmd.srcs = new_srcs
+                ir_func.num_upvalues = #cmd.srcs
+            end
         end
     end
 
-    -- 4) Propagate the constant globals, and rename the existing ones accordingly
 
-    for i, g_id in ipairs(module.exported_globals) do
-        module.exported_globals[i] = assert(new_global_id[g_id])
+    -- 4) Propagate the constants local variables and upvalues.
+
+    -- Returns a new `ir.Value` representing `src_val` after constant propagation.
+    -- @param f_data  FuncData of the function whose IR contains `src_val`.
+    -- @param func    corresponding ir.Function
+    -- @param src_val The value that we may need to update.
+    local function updated_value(f_data, func, src_val)
+        if src_val._tag == "ir.Value.LocalVar"
+           and f_data.n_writes_of_locvar[src_val.id] == 1
+           and src_val.id > #func.typ.arg_types
+           and f_data.locvar_constant_init[src_val.id] then
+
+            return f_data.locvar_constant_init[src_val.id]
```

Comment:
Style nitpick: when the condition has multiple lines, we like to put the `then` on it's own line:

```lua
if
  xxx
  and yyy
then
```